pkg/pillar/containerd/containerd.go

```diff
@@ -438,26 +378,183 @@ func getSavedImageInfo(containerID string) (v1.Image, error) {
+		}
+
+		var link netlink.Link
+		var err error
+
+		link, err = netlink.LinkByName(iface.Vif)
+		if err != nil {
+			return fmt.Errorf("Cannot find interface %s: %v", iface.Vif, err)
+		}
+
+		if err := netlink.LinkSetNsPid(link, int(pid)); err != nil {
+			return fmt.Errorf("Cannot move interface %s into namespace: %v", iface.Vif, err)
+		}
+	}
+
+	binds := []struct {
+		ns   string
+		path string
+	}{
+		{"cgroup", ""},
+		{"ipc", ""},
+		{"mnt", ""},
+		{"net", ""},
+		{"pid", ""},
+		{"user", ""},
+		{"uts", ""},
+	}
+
+	for _, b := range binds {
+		if err := bindNS(b.ns, b.path, pid); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
+// CtrInfo looks up
+func CtrInfo(name string) (int, string, error) {
+	c, err := loadContainer(name)
+	if err == nil {
+		if t, err := c.Task(ctrdCtx, nil); err == nil {
+			if stat, err := t.Status(ctrdCtx); err == nil {
+				return int(t.Pid()), string(stat.Status), nil
+			}
+		}
+	}
+	return 0, "", err
+}
+
+// CtrStart starts the default task in a pre-existing container and attaches its logging to memlogd
+func CtrStart(domainName string) (int, error) {
+	ctr, err := loadContainer(domainName)
+	if err != nil {
+		return 0, err
+	}
+
+	logger := GetLog()
+
+	// This is silly but necessary due to containerd bug
+	// https://github.com/containerd/containerd/issues/4019
+	// essentially, when you create a container and then remove it,
+	// containerd blows away everything in the parent dir of the first one it finds,
+	// in this case, "/dev/null", so it blows away everything in "/dev".
+	// This most certainly is a "bad thing".
+	//
+	// To fix it temporarily, we are creating a tmpdir and creating a null
+	// device there, so that it can blow away the tempdir
+	stdinDir := path.Join("/run", "containers-stdin", domainName)
+	if err := os.MkdirAll(stdinDir, 0700); err != nil {
+		return 0, err
+	}
+	stdinFile := path.Join(stdinDir, "null")
+	// make a dev null in stdinDir
+	if err := syscall.Mknod(stdinFile, uint32(os.FileMode(0660)|syscall.S_IFCHR), int(unix.Mkdev(1, 3))); err != nil {
+		return 0, err
+	}
+
+	io := func(id string) (cio.IO, error) {
+		stdoutFile := logger.Path(domainName + ".out")
+		stderrFile := logger.Path(domainName)
+		return &logio{
+			cio.Config{
+				Stdin:    stdinFile,
+				Stdout:   stdoutFile,
+				Stderr:   stderrFile,
+				Terminal: false,
+			},
+		}, nil
+	}
+	task, err := ctr.NewTask(ctrdCtx, io)
+	if err != nil {
+		return 0, err
+	}
+
+	if err := prepareProcess(int(task.Pid()), nil); err != nil {
+		return 0, err
+	}
+
+	if err := task.Start(ctrdCtx); err != nil {
+		return 0, err
+	}
+
+	return int(task.Pid()), nil
+}
+
+// CtrStop stops (kills) the main task in the container
+func CtrStop(containerID string, force bool) error {
+	ctr, err := CtrdClient.LoadContainer(ctrdCtx, containerID)
+	if err != nil {
+		return fmt.Errorf("can't find cotainer %s (%v)", containerID, err)
+	}
+
+	signal, err := containerd.ParseSignal(defaultSignal)
```

Comment:
Why is that function from containerd? Isn't it just an os thing to get a number from a string?

Reply:
Well, strictly speaking -- we may get by without it -- it is just a bit nicer/more flexible like that since ParseSignal parses a given string into a syscall.Signal where the rawSignal can be a string with "SIG" prefix, or a signal number in string format.

Reply:
OK